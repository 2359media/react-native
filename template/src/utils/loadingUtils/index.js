import React, {useState, useCallback} from 'react';

export const Loading = {
  show: text => {},
  hide: () => {},
};

export const LoadingContainer = ({LoadingView, children}) => {
  const [state, setState] = useState({visible: false});

  Loading.show = useCallback(text => setState({visible: true, text}), []);
  Loading.hide = useCallback(() => setState({visible: false, text: null}), []);

  return (
    <>
      <LoadingView {...state} />
      {children}
    </>
  );
};
