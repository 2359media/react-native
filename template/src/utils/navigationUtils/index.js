import 'react-native-gesture-handler';
import React from 'react';
import ModalNavigator from './ModalNavigator';
import StackNavigator from './StackNavigator';
import SwitchNavigator from './SwitchNavigator';
import TabNavigator from './TabNavigator';

const getScreen = (key, screen) => {
  if (typeof screen !== 'object') {
    return screen;
  }
  const screens = {};
  Object.keys(screen).forEach(k => (screens[k] = getScreen(k, screen[k])));

  if (key.indexOf('Switch') > 0) {
    return () => <SwitchNavigator screens={screens} />;
  } else if (key.indexOf('Tab') > 0) {
    return () => <TabNavigator screens={screens} />;
  } else if (key.indexOf('Modal') > 0) {
    return () => <ModalNavigator screens={screens} />;
  } else {
    return () => <StackNavigator screens={screens} />;
  }
};

export {NavigationContainer} from '@react-navigation/native';

export const RootNavigator = ({screens}) => {
  return getScreen('RootModal', screens)();
};
