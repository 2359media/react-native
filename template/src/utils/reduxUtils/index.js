import React from 'react';
import {createStore, combineReducers} from 'redux';
import {Provider} from 'react-redux';
import {persistStore, persistReducer} from 'redux-persist';
import storage from '@react-native-community/async-storage';
import {PersistGate} from 'redux-persist/integration/react';

export const ReduxContainer = ({
  reducers,
  initialState,
  children,
  persistKeys,
}) => {
  const combinedReducer = combineReducers(reducers);
  const persistConfig = {key: 'root', storage, whitelist: persistKeys};
  const reducer = persistReducer(persistConfig, combinedReducer);
  const store = createStore(
    reducer,
    initialState,
    (window.__REDUX_DEVTOOLS_EXTENSION__ || (() => {}))(),
  );
  const persistor = persistStore(store);

  return (
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        {children}
      </PersistGate>
    </Provider>
  );
};
