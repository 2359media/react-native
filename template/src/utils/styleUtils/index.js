export function createStylesheet(styles) {
  return styles;
}
