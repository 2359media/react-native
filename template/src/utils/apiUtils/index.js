import {useCallback} from 'react';
import {useStore} from 'react-redux';
import {Keyboard, Alert} from 'react-native';
import Axios from 'axios';
import {Loading} from '../loadingUtils';

let _baseURL;

export function useRequest() {
  const store = useStore();

  return useCallback(
    (api, {showLoading, loadingText, showError} = {}) => {
      const token = store.getState().app.token;
      const requestProps = {
        baseURL: _baseURL,
        timeout: 10000,
        headers: {'x-access-token': token},
        ...api,
      };

      showLoading && Keyboard.dismiss();
      showLoading && Loading.show(loadingText);

      return Axios.request(requestProps)
        .then(r => {
          const response = {
            data: r.data,
            status: r.status,
            statusText: (r.data || {}).message || r.statusText,
            headers: r.headers,
          };
          return response;
        })
        .catch(e => {
          e.response = e.response || {};
          const response = {
            data: e.response.data,
            status: e.response.status,
            statusText:
              (e.response.data || {}).message ||
              e.response.statusText ||
              e.message,
            headers: e.response.headers,
          };
          showError && Alert.alert('Error', response.statusText);
          throw response;
        })
        .finally(() => {
          showLoading && Loading.hide();
        });
    },
    [store],
  );
}

export const ApiContainer = ({baseURL, children}) => {
  _baseURL = baseURL;
  return children;
};
