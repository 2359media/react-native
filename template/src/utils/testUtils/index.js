import React from 'react';
import {NavigationContext} from '@react-navigation/native';
import {SafeAreaContext} from 'react-native-safe-area-context';
import TestRenderer, {act} from 'react-test-renderer';
import Axios from 'axios';
import {ReduxContainer} from '../reduxUtils';
import {Loading} from '../loadingUtils';

export const renderer = component => {
  const tree = TestRenderer.create(component);
  const get = (testID, prop) => {
    try {
      const props = tree.root.findByProps({testID}).props;
      return prop ? props[prop] : props;
    } catch {
      return undefined;
    }
  };
  const run = (id, prop = 'onPress', ...params) => {
    act(() => {
      if (typeof id === 'string') {
        get(id)[prop](...params);
      }
      jest.runAllTimers();
    });
  };
  const update = tree.update;
  const toJSON = tree.toJSON;
  return {get, run, update, toJSON};
};

export const navigation = {
  navigate: jest.fn(),
  reset: jest.fn(),
  canGoBack: jest.fn(() => true),
};

export {Alert} from 'react-native';

Loading.show = jest.fn(Loading.show);
Loading.hide = jest.fn(Loading.hide);
export {Loading};

/**
 *
 * @param {(request: any) => ({ status: number, data: any })} fn e.g. req => ({status: 200, data: {}});
 */
export const mockResponse = fn => {
  Axios.mockResponse(fn);
};

export const TestContainer = ({initialState, Component, params}) => {
  return (
    <ReduxContainer
      reducers={require('../../reducers').default}
      persistKeys={[]}
      initialState={initialState}>
      <NavigationContext.Provider value={navigation}>
        <SafeAreaContext.Provider
          value={{top: 0, bottom: 0, left: 0, right: 0}}>
          <Component route={{params}} navigation={navigation} />
        </SafeAreaContext.Provider>
      </NavigationContext.Provider>
    </ReduxContainer>
  );
};
