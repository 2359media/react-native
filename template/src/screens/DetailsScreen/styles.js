import {createStylesheet} from '../../utils/styleUtils';

const styles = createStylesheet({
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    marginBottom: 24,
  },
  body: {
    fontSize: 16,
    marginBottom: 24,
  },
});

export default styles;
