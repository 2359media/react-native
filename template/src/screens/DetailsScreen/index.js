import React from 'react';
import {Text} from 'react-native';
import {useSelector} from 'react-redux';
import Screen from '../../components/Screen';
import ContentView from '../../components/ContentView';
import TopNavigation from '../../components/TopNavigation';
import Button from '../../components/Button';
import styles from './styles';

const DetailsScreen = ({route, navigation}) => {
  const {title, body} = useSelector(s =>
    s.todo.all.find(t => t.id === route.params.id),
  );

  function gotoModalScreen() {
    navigation.navigate('Modal');
  }

  return (
    <Screen>
      <TopNavigation title="Details" />
      <ContentView>
        <Text style={styles.title}>{title}</Text>
        <Text style={styles.body}>{body}</Text>
        <Button
          testID="modalButton"
          style={styles.div}
          text="Goto modal screen"
          onPress={gotoModalScreen}
        />
      </ContentView>
    </Screen>
  );
};

export default DetailsScreen;
