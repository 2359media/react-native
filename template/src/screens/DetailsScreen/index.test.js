import React from 'react';
import {renderer, TestContainer, navigation} from '../../utils/testUtils';
import DetailsScreen from '.';

const initialState = {
  todo: {
    all: [{id: 1, title: 'title', body: 'body'}],
  },
};

const tree = renderer(
  <TestContainer
    initialState={initialState}
    params={{id: 1}}
    Component={DetailsScreen}
  />,
);

test('snapshot', () => {
  expect(tree).toMatchSnapshot();
});

test('goto modal', () => {
  tree.run('modalButton');
  expect(navigation.navigate).toBeCalledWith('Modal');
});
