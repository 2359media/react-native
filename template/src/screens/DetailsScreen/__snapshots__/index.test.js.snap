// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "white",
          "elevation": 5,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
          "shadowColor": "black",
          "shadowOpacity": 0.2,
          "shadowRadius": 5,
          "zIndex": 100,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 60,
          "paddingHorizontal": 16,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "minWidth": 50,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 8,
              "left": 8,
              "right": 8,
              "top": 8,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Image
            source={1}
            style={
              Object {
                "margin": 4,
                "tintColor": "#EC632B",
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignSelf": "stretch",
            "flex": 1,
            "justifyContent": "center",
            "paddingVertical": 8,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "black",
              "fontSize": 20,
              "fontWeight": "bold",
              "left": 50,
              "position": "absolute",
              "right": 50,
              "textAlign": "center",
            }
          }
        >
          Details
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "minWidth": 50,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Array [
        Object {
          "flexGrow": 1,
          "paddingBottom": 24,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 24,
        },
        undefined,
      ]
    }
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="handled"
  >
    <View>
      <Text
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 24,
          }
        }
      >
        title
      </Text>
      <Text
        style={
          Object {
            "fontSize": 16,
            "marginBottom": 24,
          }
        }
      >
        body
      </Text>
      <View
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 8,
            "left": 8,
            "right": 8,
            "top": 8,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#EC632B",
            "borderRadius": 8,
            "flexDirection": "row",
            "justifyContent": "center",
            "opacity": 1,
            "padding": 8,
          }
        }
        testID="modalButton"
      >
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
              "margin": 4,
            }
          }
        >
          Goto modal screen
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
